### PR TITLE
Fix snapshot reference replacement replacing parts of the snapshot keys

### PR DESCRIPTION
**Pull request by dfangl**
Supersedes #9498 

## Motivation
We regularly have the issue that the snapshot reference replacements also act upon the keys of the snapshot. This is especially annoying if it introduces test flakyness due to randomized variable names - if the name per chance matches up with a part of the key name, it will be replaced.
This is especially true for short resource names, which are likely to be present in the snapshot key name.

## Changes
The reference replacement never replaces the top level snapshot key - only everything in the actual snapshot values.

### CODE REVIEW

File: tests/aws/services/lambda_/test_lambda.snapshot.json [hunk 1]
```diff
@@ -3456,15 +3456,15 @@
     }
   },
   "tests/aws/services/lambda_/test_lambda.py::TestLambdaBaseFeatures::test_assume_role[1]": {
-    "recorded-date": "10-10-2023, 14:55:53",
+    "recorded-date": "07-11-2023, 17:06:55",
     "recorded-content": {
-      "invoke-result-assumed-role-arn": "arn:aws:sts::111111111111:assumed-role/lambda-autogenerated-<lambda-autogenerated-role-prefix>/<resource:1>"
+      "invoke-result-assumed-role-arn": "arn:aws:sts::869636330914:assumed-role/lambda-autogenerated-<lambda-autogenerated-role-prefix>/<function-name>@lambda_function"
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda.snapshot.json:3461]: I guess we still need a transformer for the account here after clearing the default transformers 🤔

Comment (dominikschubert) [inline tests/aws/services/lambda_/test_lambda.snapshot.json:3461]:
```suggestion
      "invoke-result-assumed-role-arn": "arn:aws:sts::111111111111:assumed-role/lambda-autogenerated-<lambda-autogenerated-role-prefix>/<function-name>@lambda_function"
```

the clear operation also removed the account id replacement

Comment (dfangl) [reply tests/aws/services/lambda_/test_lambda.snapshot.json]: Yes, totally missed that one. Corrected!

File: tests/aws/services/lambda_/test_lambda.py [hunk 2]
```diff
@@ -276,8 +278,14 @@ def test_assume_role(self, create_lambda_function, aws_client, snapshot, functio
...
         """
         lambda_autogen_role = "(?<=lambda-autogenerated-)([0-9a-f]{8})"
+        # avoid any other transformers possible registering reference replacements (especially resource transformer)
+        snapshot.transformers.clear()
         snapshot.add_transformer(
-            snapshot.transform.regex(lambda_autogen_role, "<lambda-autogenerated-role-prefix>")
+            [
+                snapshot.transform.regex(account_id, "1" * 12),
+                snapshot.transform.regex(lambda_autogen_role, "<lambda-autogenerated-role-prefix>"),
+                snapshot.transform.regex(r'(?<=/)[a-zA-Z]{1,2}(?="|@)', "<function-name>"),
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda.py:287]: much more elegant than the hacky conditional assertions 🥇 👏👏👏 

We can get rid of the `if function_name_length == 1:` conditional at the end of the test now.

**Review by steffyP — APPROVED** [review]
Thanks for the fix! 🙏

**Pull request merged**